File: Linked List/61.rotate-list.cpp
/*
 * @lc app=leetcode id=61 lang=cpp
 *
 * [61] Rotate List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        auto first = head, second = head;
        int n = 1;
        if(!head) return head;
        for(auto p = head; p->next; p = p->next)
            n++;
        k %= n;
        while(k--)
            first = first->next;
        while(first->next)
        {
            second = second->next;
            first = first->next;
        }
        first->next = head;
        head = second->next;
        second->next = NULL;
        return head;
    }
};
// @lc code=end

File: Linked List/61.rotate-list.py
#
# @lc app=leetcode id=61 lang=python3
#
# [61] Rotate List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        cur, first, second = head, head, head
        n = 1
        if not head:
            return head
        while cur.next:
            n += 1
            cur = cur.next
        k %= n
        while k:
            first = first.next
            k -= 1
        while first.next:
            first = first.next
            second = second.next
        first.next = head
        head = second.next
        second.next = None
        return head
# @lc code=end

